lib/job/compute-next-run-at.ts

```diff
@@ -34,40 +34,68 @@ export const computeNextRunAt = function(this: Job) {
+        cronOptions.currentDate = new Date(lastRun.getTime() + 1000);
         cronTime = parser.parseExpression(interval, cronOptions);
         nextDate = cronTime.next().toDate();
       }
 
+      // If start date is present, check if the nextDate should be larger or equal to startDate. If not set startDate as nextDate
+      if (startDate !== null) {
+        startDate = moment.tz(moment(startDate).format('YYYY-MM-DD'), timezone).toDate();
+        if (startDate > nextDate) {
+          cronOptions.currentDate = startDate;
+          cronTime = parser.parseExpression(interval, cronOptions);
+          nextDate = cronTime.next().toDate();
+        }
+      }
+
+      // If job has run in the past and skipDays is not null, add skipDays to nextDate
+      if ((dateNow > lastRun) && skipDays !== null) {
+        try {
+          nextDate = new Date(nextDate.getTime() + (humanInterval(skipDays) ?? 0));
+        } catch {}
+      }
+
+      // If endDate is less than the nextDate, set nextDate to null to stop the job from running further
+      if (endDate !== null) {
+        const endDateDate: Date = moment.tz(moment(endDate).format('YYYY-MM-DD'), timezone).toDate();
+        if (nextDate > endDateDate) {
+          nextDate = null;
+        }
+      }
+
       this.attrs.nextRunAt = nextDate;
-      debug('[%s:%s] nextRunAt set to [%s]', this.attrs.name, this.attrs._id, new Date(this.attrs.nextRunAt).toISOString());
+      debug('[%s:%s] nextRunAt set to [%s]', this.attrs.name, this.attrs._id, this.attrs.nextRunAt.toISOString());
     // Either `xo` linter or Node.js 8 stumble on this line if it isn't just ignored
     } catch {
+      debug('[%s:%s] failed nextRunAt based on interval [%s]', this.attrs.name, this.attrs._id, interval);
       // Nope, humanInterval then!
       try {
         if (!this.attrs.lastRunAt && humanInterval(interval)) {
-          this.attrs.nextRunAt = lastRun.valueOf();
-          debug('[%s:%s] nextRunAt set to [%s]', this.attrs.name, this.attrs._id, new Date(this.attrs.nextRunAt).toISOString());
+          this.attrs.nextRunAt = lastRun;
+          debug('[%s:%s] nextRunAt set to [%s]', this.attrs.name, this.attrs._id, this.attrs.nextRunAt.toISOString());
         } else {
-          this.attrs.nextRunAt = (lastRun as moment.Moment).valueOf() + (humanInterval(interval) ?? 0);
-          debug('[%s:%s] nextRunAt set to [%s]', this.attrs.name, this.attrs._id, new Date(this.attrs.nextRunAt).toISOString());
+          this.attrs.nextRunAt = new Date(lastRun.getTime() + (humanInterval(interval) ?? 0));
+          debug('[%s:%s] nextRunAt set to [%s]', this.attrs.name, this.attrs._id, this.attrs.nextRunAt.toISOString());
         }
       // Either `xo` linter or Node.js 8 stumble on this line if it isn't just ignored
       } catch {}
     } finally {
-      if (Number.isNaN(this.attrs.nextRunAt)) {
+      if (Number.isNaN(this.attrs.nextRunAt.getTime())) {
```

Comment:
i think this check was for checking if "date" is valid. now if nextRunAt is not set (which can be the case as far as I remeber) it could break.. maybe we can fix this in an upcoming PR when you added the types for attrs (and make nextRunAt optional, typescript should complain)

Reply:
@simllll  thanks. The 'date' is valid check is still needed here. There are 3-4 test-cases testing this flow.
In the old code `this.attrs.nextRunAt` was `this.attrs.nextRunAt.getTime()`.
`this.attrs.nextRunAt` is "Invalid Date".
To test if a date is invalid, you have to use `isNan(date.getTime())`

https://stackoverflow.com/questions/1353684/detecting-an-invalid-date-date-instance-in-javascript

Reply:
I see your point, i just thought that nextrunat could be undefined already,which will let the code fail if you access a property/method on "undefined". But when it was the case before, its probably good to go (y)